CET save issue for CrudListenerScripts during upgradation of  module installed .

**issue with CET's CrudListenerScripts**
if we add a new CET and add a new CrudListenerScript for it, it is not possible to upgrade the module installed on another meveo instance

**To Reproduce**
After install a meveo module in two meveo instance  then add a new CET with add a new CrudListenerScript  and  push the changes to github     then  pull from the other meveo instance , it will generate this error.

**Error Log :**

2023-07-04 01:48:16,967 ERROR [org.jboss.as.ejb3.invocation] (default task-1) WFLYEJB0034: EJB Invocation failed on component CustomEntityTemplateApi for 
method public org.meveo.model.IEntity org.meveo.api.CustomEntityTemplateApi.createOrUpdate(org.meveo.api.dto.BaseEntityDto) 
throws org.meveo.api.exception.MeveoApiException,org.meveo.admin.exception.BusinessException:
javax.ejb.EJBTransactionRolledbackException: Script com.unikbase.rfid.NFCTagListenerScript does not exists
	at org.jboss.as.ejb3@18.0.0.Final//org.jboss.as.ejb3.tx.CMTTxInterceptor.invokeInCallerTx(CMTTxInterceptor.java:203)
Caused by: java.lang.IllegalArgumentException: Script com.unikbase.rfid.NFCTagListenerScript does not exists
	at deployment.meveo.war//org.meveo.api.CustomEntityTemplateApi.fromDTO(CustomEntityTemplateApi.java:648).

**Root Casue**

when prepare CustomEntityTemplate object  on that time need to initialize --crudEventListenerScript attribute , 
it is depend on a scriptinstance .but this dependent object is not persist already .  
for this reason it gives giving null value. 



![image](https://github.com/meveo-org/meveo/assets/9381810/66489a58-07e3-4ec6-ab08-f899da936724)

**Solution**
Need to  update a CET  if (cetDto.getCrudEventListenerScript() != null)  after complete installation of  all module item .

--- a/meveo-api/src/main/java/org/meveo/api/CustomEntityTemplateApi.java
+++ b/meveo-api/src/main/java/org/meveo/api/CustomEntityTemplateApi.java
@@ -643,9 +643,8 @@ public class CustomEntityTemplateApi extends BaseCrudApi<CustomEntityTemplate, C
         
         if(dto.getCrudEventListenerScript() != null) {
         	var crudListenerScript = scriptInstanceService.findByCode(dto.getCrudEventListenerScript());
-        	cet.setCrudEventListenerScript(crudListenerScript);
-        	if(crudListenerScript == null) {
-        		throw new IllegalArgumentException("Script " + dto.getCrudEventListenerScript() + " does not exists");
+        	if(crudListenerScript != null) {
+        		cet.setCrudEventListenerScript(crudListenerScript);
         	}
         }
 

--- a/meveo-api/src/main/java/org/meveo/api/module/MeveoModuleApi.java
+++ b/meveo-api/src/main/java/org/meveo/api/module/MeveoModuleApi.java
@@ -1761,6 +1761,11 @@ public class MeveoModuleApi extends BaseCrudApi<MeveoModule, MeveoModuleDto> {
 			meveoModuleItemInstaller.uninstallItem(options, null, item);
 		}
 		
+		// update CET
+		for (var itemDto : meveoModuleItemInstaller.getSortedModuleItems(installItems)) {
+		  meveoModuleItemInstaller.updateCetforCrudEventListenerScript(itemDto);
+		}
+		
 		moduleCtx.end();
 		
 		meveoModuleService.update(module);

--- a/meveo-api/src/main/java/org/meveo/api/module/MeveoModuleItemInstaller.java
+++ b/meveo-api/src/main/java/org/meveo/api/module/MeveoModuleItemInstaller.java
@@ -818,6 +818,37 @@ public class MeveoModuleItemInstaller {
 			}
 		}
 	}
+	
+	public void updateCetforCrudEventListenerScript (MeveoModuleItemDto moduleItemDto) throws BusinessException {
+
+		if (moduleItemDto.getDtoClassName().equals(CustomEntityTemplateDto.class.getName())) {
+				CustomEntityTemplateDto cetDto =null;
+				if(moduleItemDto.getDtoData() instanceof CustomEntityTemplateDto){
+				cetDto = (CustomEntityTemplateDto) moduleItemDto.getDtoData();
+				} else {
+					try {
+						Class<? extends BaseEntityDto> dtoClass = (Class<? extends BaseEntityDto>) Class.forName(moduleItemDto.getDtoClassName());
+						cetDto = (CustomEntityTemplateDto) JacksonUtil.convert(moduleItemDto.getDtoData(), dtoClass);
+					} catch (ClassNotFoundException e) {
+						log.error("Cannot find dto class", e);
+						throw new BusinessException("cannot find crusdScript class "+moduleItemDto.getDtoClassName());
+					}
+				}
+				if (cetDto.getCrudEventListenerScript() != null) {
+					CustomEntityTemplate cet = customEntityTemplateService.findByCode(cetDto.getCode());
+					ScriptInstance si = scriptInstanceService.findByCode(cetDto.getCrudEventListenerScript());
+					if (si != null) {
+						cet.setCrudEventListenerScript(si);
+						customEntityTemplateService.update(cet);
+					} else {
+						throw new BusinessException("cannot find crusdScript "+cetDto.getTransientCrudEventListenerScript()+" of entity "+cetDto.getCode());
+					}
+				}
+			
+		}
+	
+}
+
 
 
 }

